src/reuse/lint.py

```diff
@@ -8,329 +8,260 @@
 """
 
 import contextlib
+import json
 import os
 import sys
 from gettext import gettext as _
-from typing import Iterable
+from pathlib import PosixPath
 
-from . import __REUSE_version__
 from .project import Project
 from .report import ProjectReport
 
 
-def _write_element(element, out=sys.stdout):
-    out.write("* ")
-    out.write(str(element))
-    out.write("\n")
-
-
-def lint(report: ProjectReport, out=sys.stdout) -> bool:
-    """Lint the entire project."""
-    bad_licenses_result = lint_bad_licenses(report, out)
-    deprecated_result = lint_deprecated_licenses(report, out)
-    extensionless = lint_licenses_without_extension(report, out)
-    missing_licenses_result = lint_missing_licenses(report, out)
-    unused_licenses_result = lint_unused_licenses(report, out)
-    read_errors_result = lint_read_errors(report, out)
-    files_without_cali = lint_files_without_copyright_and_licensing(report, out)
-
-    lint_summary(report, out=out)
-
-    success = not any(
-        any(result)
-        for result in (
-            bad_licenses_result,
-            deprecated_result,
-            extensionless,
-            missing_licenses_result,
-            unused_licenses_result,
-            read_errors_result,
-            files_without_cali,
-        )
+def add_arguments(parser):
+    """Add arguments to parser."""
+    parser.add_argument(
+        "-q", "--quiet", action="store_true", help="Prevents output"
     )
+    mutex_group = parser.add_mutually_exclusive_group()
+    mutex_group.add_argument(
+        "-j", "--json", action="store_true", help="Formats output as JSON"
+    )
+    mutex_group.add_argument(
+        "-p",
+        "--plain",
+        action="store_true",
+        help="Formats output as plain text",
+    )
+    mutex_group.add_argument(
+        "--format",
+        nargs="?",
+        choices=("json", "plain"),
+        help="Formats output using the chosen formatter",
+    )
+
 
-    out.write("\n")
-    if success:
-        out.write(
-            _(
-                "Congratulations! Your project is compliant with version"
-                " {} of the REUSE Specification :-)"
-            ).format(__REUSE_version__)
+# pylint: disable=too-many-locals, too-many-branches, too-many-statements
+def format_plain(report: ProjectReport) -> str:
+    """Formats data dictionary as plaintext string to be printed to sys.stdout
+
+    :param report: ProjectReport data
+    :return: String (in plaintext) that can be output to sys.stdout
+    """
+    output = ""
+    data = report.to_dict()
+
+    # If the project is not compliant:
+    if not data["summary"]["compliant"]:
+        # Bad licenses
+        bad_licenses = data["non_compliant"]["bad_licenses"]
+        if bad_licenses:
+            output += "# " + _("BAD LICENSES") + "\n\n"
```

Comment:
`output +=` is really bad for performance, especially for large strings (which this might well be), because a new string object is created every time (Python strings are immutable). Output should probably be a StringIO object (which is append-only mutable, afaik) that we convert to a string at the end.

Reply:
Done :)